Update/storybook commands to use wireit

louwie17 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This does two things:
- Introduce the wireit commands, which should come with some caching benefits
- Add WooCommerce Blocks to the watch command so it is also shown in our main storybook when running watch locally. ( cc: @opr the introduction of `--workspace-concurrently` made this a lot easier )

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Build storybook using the env variable, like so: `STORYBOOK_COMPOSITION_PATH_PREFIX="storybook-static" pnpm --filter=@woocommerce/storybook run build`
2. Serve the `storybook` directory with `http-server` -> `npx http-server tools/storybook` ( that way we simulate the folder structure as storybook lives within `storybook-static`
3. Go to the url and click on the `storybook-static` folder, this should load storybook and load WooCommerce Blocks correctly as well.
4. Do the above again without the environment variable and server the `storybook-static` folder directly: `npx http-server tools/storybook/storybook-static`
5. Stories should still load correctly including the WooCommerce Blocks ( you may have to do a hard refresh so the browser clears its cache )
6. Run the watch command: `pnpm --filter=@woocommerce/storybook run watch:build` and go to `http://localhost:6007/`
7. It should load storybook and also render the WooCommerce Block story book correctly
8. Go to `http://localhost:6006/` this should render WooCommerce Block storybooks only.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

louwie17 committed:
Integrated wireit to storybook

louwie17 committed:
Attempt at integrating wireit

louwie17 committed:
Update wireit dependencies for storybook package

louwie17 committed:
Add changelog

louwie17 commented on tools/storybook/.storybook/main.js:
```diff
@@ -26,7 +26,14 @@ module.exports = {
+				'woocommerce-blocks': {
+					expanded: false,
+					title: 'WooCommerce Blocks',
+					url: 'http://localhost:6006',
```
With the `workspace-concurrency` command blocks storybook is also being run when running `watch:build` on the storybook. This allows us to show it within the main storybook just like on Github Pages.

louwie17 commented on tools/storybook/package.json:
```diff
@@ -11,13 +11,15 @@
-		"preinstall": "npx only-allow pnpm",
-		"storybook": "pnpm build-woocommerce && ./import-wp-css-storybook.sh && BABEL_ENV=storybook STORYBOOK=true storybook dev -c ./.storybook -p 6007 --ci",
-		"storybook-rtl": "USE_RTL_STYLE=true pnpm storybook"
+		"build": "pnpm --if-present --workspace-concurrency=Infinity --stream --filter=\"$npm_package_name...\" '/^build:project:storybook$/'",
```
Only trigger the `build:project:storybook` as we don't actually need to build the other dependencies, storybook does that for us.

ObliviousHarmony replied:
Rather than only building `build:project:storybook` (the regex is unnecessary here too), we should remove the `--filter` option that way it will only run on the current package. With that done, we can also remove `--if-present` since we _know_ the script is present here.

louwie17 commented:
@ObliviousHarmony or @opr would I be able to get a review from one of you? Christopher because of the wireit stuff, and Thomas because this integrates blocks storybook on dev as well ( as discussed in a previous PR ).

opr reviewed (approved):
Thanks @louwie17 this is working for me according to testing instructions. I think @ObliviousHarmony should review too for additional confidence, since I'm not familiar with the build steps here.

ObliviousHarmony commented on plugins/woocommerce-blocks/package.json:
```diff
@@ -74,9 +74,11 @@
 		"start": "rimraf build/* && cross-env BABEL_ENV=default CHECK_CIRCULAR_DEPS=true webpack --watch",
-		"storybook": "storybook dev  -c ./storybook -p 6006 --ci",
-		"storybook:build": "BABEL_ENV=development storybook build  -c ./storybook -o ./storybook/dist",
+		"storybook": "pnpm --if-present '/^watch:build:project:storybook$/'",
+		"storybook:build": "pnpm --if-present '/^build:project:storybook$/'",
```
I'm _pretty sure_ that the regular expression here doesn't do anything anymore, no? The `--if-present` option also isn't necessary since it's just calling another script. Just `pnpm watch:build:project:storybook` is sufficient.

louwie17 replied:
Sounds good, removed as part of: https://github.com/woocommerce/woocommerce/pull/51701/commits/6489e5f0ce19371b785d7c2574cd06c5e8bc5b28

ObliviousHarmony commented on tools/storybook/package.json:
```diff
@@ -48,12 +50,71 @@
+			"output": [
+				"storybook-static"
+			],
+			"dependencies": [
+				"import-wp-css-for-storybook",
+				"dependencyOutputs",
+				"../../plugins/woocommerce-blocks:build:project:storybook"
+			]
```
While `wireit` _does_ support script dependencies, thus far, we've only used them to share file input references. When you use `dependencies` like this it has `wireit` manage the task orchestration rather than `pnpm`. This should probably just be part of the script.

louwie17 replied:
Thanks, turns out it wasn't necessary anyway as the `pnpm` script already ran build command command of `woocommerce-blocks`. Removed it here: https://github.com/woocommerce/woocommerce/pull/51701/commits/9e04d2312428033bbf35b69dd9396f238c982210

ObliviousHarmony reviewed (commented):
Sorry for the wait @louwie17!

I left some comments and I also wanted to share [this documentation from a draft PR](https://github.com/woocommerce/woocommerce/blob/34e6d4e7d3c28fa99284ef951f1b592993cc79fc/tools/README.md) (#51820). This might shed a bit more light on how all of this works. I'd _love_ if you could provide feedback there too so that I can make sure it's fleshed out sufficiently!

louwie17 committed:
Remove unneeded --if-present and regex in pnpm command

louwie17 committed:
Remove unneeded dependency for storybook build command

louwie17 commented:
> I left some comments and I also wanted to share [this documentation from a draft PR](https://github.com/woocommerce/woocommerce/blob/34e6d4e7d3c28fa99284ef951f1b592993cc79fc/tools/README.md) (#51820). This might shed a bit more light on how all of this works. I'd _love_ if you could provide feedback there too so that I can make sure it's fleshed out sufficiently!

Thanks for the review @ObliviousHarmony, I addressed your comments. I also left two comments on your documentation PR with some suggestions. Thanks for working on that 👍

ObliviousHarmony commented on plugins/woocommerce-blocks/package.json:
```diff
@@ -391,10 +393,30 @@
 				"dependencyOutputs"
 			]
 		},
+		"build:project:storybook": {
```
Sorry, I guess I didn't notice before but I think this is missing the `dependencyOutputs` dependency?

louwie17 replied:
Addressed: https://github.com/woocommerce/woocommerce/pull/51701/commits/941fbccdfa41c075e60449bab363f93c08265312

louwie17 committed:
Add missing dependency in blocks storybook build script

ObliviousHarmony reviewed (approved):
Great work :smile:
